src/sim.jl

```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env julia
+using Distributions
+using JuliaDB
+using Unitful
+
+include("RoutingKit.jl")
+
+
+
+"""
+For a given trip_id, determine how much time is spent waiting at stops during 
+that trip.
+
+Args:
+  stop_prob  - Probability of stopping at a given stop with zero duration
+  zstop_time - Seconds spent at a stop that is otherwise given zero duration
+"""
+function GetStopChargingTime(stops, trip_id; stop_prob=0.3, zstop_time=10)
+    #Get list of stops used by this trip and merge in the stop data
+    trip_stops = filter(row->row[:trip_id]==trip_id, stops)
+    trip_stops = filter(row->row[:inductive_charging]==true, trip_stops)
+    #Total time we are scheduled to be stopped
+    stopped_time = sum(trip_stops.stop_duration)
+
+    #Some of the stops with zero duration we do end up stopping at briefly
+    zero_count = sum(trip_stops.stop_duration.==0)     #Number of zero-duration stops
+    zero_count = rand(Binomial(zero_count, stop_prob)) #Number we actually stop at
+    stopped_time += zero_count*zstop_time              #Add time spent at them
+
+    return stopped_time
+end
+
+
+
+"""
+Determine <travel time (s), travel distance (m)> between two latlng points using
+the road network held in router.
+
+Args:
+    search_radius - Snap stop to closest road network node searching within
+                    this radius.
+"""
+function TimeDistanceBetweenPoints(router, ll1, ll2; search_radius=1000)
+    try
+        RoutingKit.getTravelTime(router, ll1[:lat], ll1[:lng], ll2[:lat], ll2[:lng], 1000)
+        return (time=time_dist[1]u"s", dist=time_dist[2]u"m")
+    catch
+        println("Warning: Couldn't find a node near either ",ll1," or ",ll2)
+        return (time=15u"minute", dist=5u"km") 
+    end
+end
+
+
+
+"""
+Simulate bus movement along a route. Information about the journey is stored by
+modifying "block".
+
+Args:
+    block   - Block of trips to be simulated (MODIFIED WITH OUTPUT)
+    stop_ll - Dictionary containing the latitudes and longitudes of stops
+    router  - Router object used to determine network distances between stops
+    depots  - List of depot locations
+    params  - Model parameters
+"""
+function RunBlock(block, stop_ll, router, depots, params)
```

Comment:
If I am reading this correctly, this function still suffers from a bug I called out in #24, which will cause a inf loop. This could easily be present for long routes, esp for feeds that describe intercity bus networks such as Greyhound.

Reply:
should be a quick check on the `if` statement that begins in 97?

Reply:
This won't cause an infinite loop: the bus will just return to its end depot with negative energy availability, which indicates an error condition. See Line 105:
```
            if energy_from_start_to_depot < prevtrip.energy_left
                println("\nEnergy trap found!")
            end
```